tests/test_package.py

```diff
@@ -163,95 +163,100 @@ def test_package_safe_name_is_correct(pkg_name, expected_name):
     package = package_file.PackageFile(
         filename="tests/fixtures/deprecated-pypirc",
         comment=None,
-        metadata=pretend.stub(name=pkg_name),
+        metadata=dict(name=pkg_name, version="1.2.3"),
         python_version=None,
         filetype=None,
     )
 
     assert package.safe_name == expected_name
 
 
-def test_metadata_dictionary_keys():
-    """Merge multiple sources of metadata into a single dictionary."""
-    package = package_file.PackageFile.from_filename(helpers.SDIST_FIXTURE, None)
-    assert set(package.metadata_dictionary()) == {
-        # identify release
+def test_supported_metadata_fields():
+    """Check that ``twine.package.PackageMetadata`` contains the expected fields."""
+    assert set(package_file.PackageMetadata.__annotations__) == {
```

Comment:
Relying on `__annotations__` breaks what is being tested here. The test here is that we serialize this to a dictionary for sending through the HTTP client as a `multipart/form-data` upload. Not that a class has the right fields.

Reply:
My reading of the test is that it just checked the dictionary fields. The serialization to `multipart/form-data` happens at a different layer and before that the metadata goes through transformation in `twine.repository.Repository._convert_data_to_list_of_tuples()` (now `twine.repository.Repository._convert_metadata_to_list_of_tuples()`). This just checked that a dictionary with the right keys is instantiated. It relied on dictionary entries were inserted with `None` value for metadata fields without a value (which would be invalid if `warehouse` were not to have explicit support for it to accommodate `twine`, BTW). Now fields that are not present in the package metadata are not populated and `packaging` is responsible for metadata validation. I decided to keep the test to verify that `packaging` does not rename, removes, or add dictionary entries.

I added a test in `test_repository.py` that tests something closer to the `multipart/form-data` format, but it ides not check the filed names. If we think that checking the field names is important, I can add it there.

Reply:
Also, the `test_metadata_dictionary_values()` also tests the metadata fields, but for how it is structured, it cannot test for the presence of fields unknown to `twine`. I think the two tests together cover all is relevant to test.

Reply:
@dnicolodi so to be clear, this isn't testing the `multipart/form-data` encoding. This is testing that the fields we know are valid and are passed through, are placed into the dictionary that is then used to generate the `multipart/form-data` encoding. Again, `__annotations__` is not the same as testing the supported metadata fields.

To be clear, there are different things here:

1. You're testing a TypedDict here that is inheriting from a dict we do not control. We don't run tests frequently enough to detect a problem with `packaging` introduced somewhere before `packaging` or `twine` releases so we risk breaking others by having a non-pinned `packaging` version.
2. The libraries we're using (and which most other libraries mimic) are then expecting  a dictionary. We don't need to test those libraries and that they produce the proper serialized data, but instead that what we pass to them is consistent. That was the purpose of this test. I _know_ because _I_ wrote it.
3. You're test in `test_repository` isn't testing `multipart/form-data` encoding either. It's testing how we prefer to pass it into there such that items are vaguely ordered. You're also effectively testing a private method which doesn't seem valuable it just makes it harder to refactor.

Reply:
You're missing the point and either GitHub lost a comment with your changes or something else happened (where it didn't properly post) but our PackageMetadata object should _not_ inherit from packaging's RawMetadata class. That's a very fragile assumption to make about the stability/implementation of that class. If this were done well, we'd use it as a collaborator, not as a parent. It's not a parent, we don't own it. As a result, your test change here doesn't make sense.

Reply:
> If this were done well ... your test change here doesn't make sense

I would like you to use a more friendly tone. I am not your subordinate.